### PR TITLE
ROX-21883: Add Scanner V4 to certgen API

### PR DESCRIPTION
**Pull request by dcaravel**
## Description

Adds the implementation for `/api/extensions/certgen/scanner?v=4` that when invoked will issue/download Scanner V4 TLS secrets, specifically for:
- `scanner-v4-indexer-tls`
- `scanner-v4-matcher-tls`
- `scanner-v4-db-tls`

A query string was chosen over a new path element to make removal easier in the future, such as when only a single scanner exists again, at which point the implementation can be changed to ignore the query string.

This API will be called when clicking the `Download YAML` button on the System Health page and expiration banners for Scanner V4 (System health UI changes implemented in https://github.com/stackrox/stackrox/pull/9551) 

image

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] ~Evaluated and added CHANGELOG entry if required~
- [x] ~Determined and documented upgrade steps~
- [x] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Unit tests and via downloading / applying the yamls produced by the new app as follows:

Downloaded the yamls via curl and verified have the desired secrets:

```
$ curl -ksS -H "Authorization: Bearer $TOKEN" -X POST "https://$ACS_HOST/api/extensions/certgen/scanner?v=4" -o k8s-secrets.yaml

$ cat k8s-secrets.yaml | yq '.metadata.name'
scanner-v4-indexer-tls
---
scanner-v4-matcher-tls
---
scanner-v4-db-tls
```

Viewed the current cert expiration date from Scanner V4:
```
$ curl -ksS -H "Authorization: Bearer $TOKEN" "https://$ACS_HOST/v1/credentialexpiry?component=SCANNER_V4" | jq
{
  "expiry": "2025-01-29T17:24:00Z"
}
```

Applied the new secrets, and restarted Scanner V4 DB, Indexer, and Matcher to pick up the new secrets, then verified the expiry date has been updated:

```
$ k apply -f k8s-secrets.yaml 
secret/scanner-v4-indexer-tls configured
secret/scanner-v4-matcher-tls configured
secret/scanner-v4-db-tls configured

$ k delete pods -lapp=scanner-v4-db
$ k delete pods -lapp=scanner-v4-indexer
$ k delete pods -lapp=scanner-v4-matcher

$ curl -ksS -H "Authorization: Bearer $TOKEN" "https://$ACS_HOST/v1/credentialexpiry?component=SCANNER_V4" | jq
{
  "expiry": "2025-01-29T21:35:00Z"
}
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: central/certgen/central_scanner.go [hunk 1]
```diff
@@ -102,6 +107,24 @@ func (s *serviceImpl) scannerHandler(w http.ResponseWriter, r *http.Request) {
...
+		httputil.WriteGRPCStyleError(w, codes.Internal, err)
+		return
+	}
+
+	rendered, err := renderer.RenderScannerV4TLSSecretOnly(renderer.Config{
+		K8sConfig:      &renderer.K8sConfig{},
+		SecretsByteMap: secrets,
+	}, defaults.GetImageFlavorFromEnv())
+	if err != nil {
+		httputil.WriteGRPCStyleErrorf(w, codes.Internal, "failed to render scanner TLS file: %v", err)
```
Comment (RTann) [inline central/certgen/central_scanner.go:121]: v4

File: pkg/renderer/kubernetes.go [hunk 2]
```diff
@@ -38,6 +39,8 @@ const (
 	scannerTLSOnly
 	// centralDBOnly renders only the central db
 	centralDBOnly
+	// scannerV4TLSOnly renders only the scanner v4 tls secrets.
+	scannerV4TLSOnly
```
Comment (RTann) [inline pkg/renderer/kubernetes.go:43]: unrelated, but I guess we won't have a mode that allows Scanner v4 only?

File: pkg/renderer/kubernetes.go [hunk 3]
```diff
@@ -132,15 +135,36 @@ func RenderCentralDBOnly(c Config, imageFlavor defaults.ImageFlavor) ([]*zip.Fil
...
+	if len(files) != numFiles {
+		return nil, utils.ShouldErr(errors.Errorf("expected %d but got %d files when rendering in mode %s", numFiles, len(files), mode))
+	}
+
+	if len(files) == 1 {
+		// Short circuit if a single file.
+		return files[0].Content, nil
+	}
+
+	buf := bytes.Buffer{}
```
Comment (RTann) [inline pkg/renderer/kubernetes.go:156]: nit: `var buf bytes.Buffer`

**Review by RTann — APPROVED** [review]
LGTM but take a look at the minor comments I left

File: pkg/renderer/kubernetes.go (hunk 2 above)
Comment (dcaravel) [reply pkg/renderer/kubernetes.go:43]: Don't believe thats needed unless we want to create a separate `roxctl scanner generate` command specific for scanner v4.  https://github.com/stackrox/stackrox/pull/9619 adjusted the output of `roxctl scanner generate` to produce both the scanner v2 and v4 files.

**Comment by dcaravel** [comment]
CI failures are unrelated to change, and are being address in https://github.com/stackrox/stackrox/pull/9653

image

**Pull request merged**